### DISCUSSION TITLE
FastAPI code generator generating unhashable models when using an enum

### DISCUSSION THREAD

**Discussion by ChrisDAT20** [fastapi/fastapi#11814] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
* In an OpenAPI schema definition file, I defined an enum type:

DraftType:
  type: string
  enum:
	- Type1
	- Type2
	- Type3
```
which is generated to the following Python class:
```
class DraftType(str, Enum):
    Type1 = "Type1"
    Type2 = "Type2"
    Type3 = "Type3"
```
It is used in another type:
```
Draft:
  type: object
  properties:
	draftType:
	  $ref: "#/components/schemas/DraftType"
	version:
	  type: integer
	  format: int64
  required:
	- draftType
	- version
```
The classes are then used in the FastAPI app:
```
@async_http_exception
async def get_draft(
    draft_request: DraftRequest = Body(None, description=""),
) -> None:
    draft: Draft = DraftLoader.get_draft(draft_request=draft_request)
```

```
class DraftRequest(BaseModel):

    draft_type: DraftType = Field(alias="draftType")

DraftRequest.update_forward_refs()
```

```
class DraftLoader:
	...
	
	@staticmethod
	@lru_cache()
	def get_draft(draft_request: DraftRequest) -> Draft:
		...
		# fetch latest draft version...
		draft: Draft = ...
		...
		return draft

```

However, the code raises an error when calling `DraftLoader.get_draft`:
```
E             File "<!- some file path ->", line 46, in get_draft
E               draft: Draft = DraftLoader.get_draft(draft_request=draft_request)
E           TypeError: unhashable type: 'DraftRequest'
```
```


### Description

As the code is generated by FastAPI, I do not have an idea on how to make the enum hashable. 
Is there something I can add to make the enums hashable?

### Operating System

Linux

### Operating System Details

_No response_

### FastAPI Version

0.110.2

### Pydantic Version

1.10.15

### Python Version

Python 3.9.13

### Additional Context

_No response_

**Comment by YuriiMotov** [comment]
That's because you are using `lru_cache` and it has the following constraint:

> Since a dictionary is used to cache results, the positional and keyword arguments to the function must be [hashable](https://docs.python.org/3/glossary.html#term-hashable).


To solve this you need to implement the `__hash__` and `__eq__` methods for `DraftRequest`.

Something like this (I haven't run the code below, it can contain mistakes):
```
class DraftRequest(BaseModel):
    draft_type: DraftType = Field(alias="draftType")

    def __hash__(self):
        return hash(self.draft_type.value)

    def __eq__(self, other: Any):
        if not isinstance(other, DraftRequest):
            return self.draft_type == other.draft_type
        return False
```

**Reply by ChrisDAT20** [reply]
I understood that hash and eq functions are needed. But how can I make FastAPI to add the functions to the generated code? I would like to avoid manual code changes to make the enums hashable.

**Reply by YuriiMotov** [reply]
Oh, I see.
I think it would be more correctly to say `code is generated by FastAPI code generator`, not `code is generated by FastAPI`.
Do you use this project: https://github.com/koxudaxi/fastapi-code-generator ?

**Reply by ChrisDAT20** [reply]
yes, you are right. Its `fast-code-generator` installed using pip.

**Reply by YuriiMotov** [reply, marked as answer]
Why don't you ask this question in the repo of `fast-code-generator`?
I think it's closer to that project then to FastAPI itself..

**Reply by YuriiMotov** [reply]
You can try using `lru_cache` from `methodtools` (`from methodtools import lru_cache`) (https://github.com/youknowone/methodtools)

**Reply by ChrisDAT20** [reply]
For some reason I thought fastapi and fastapi-code-generator are maintained together. I posted the issue there as well.
Thank you for the hint!
